### DISCUSSION TITLE
Method Map.prototype.set called on incompatible receiver #<Map> when running MikroORM.init() in global jest setup

### DISCUSSION THREAD

**Discussion by kpervin** [mikro-orm/mikro-orm#3795] [Q&A]
**Describe the bug**
When adding a basic global setup script to jest, the global setup file will run fine:

```ts
import config from "../src/mikro-orm.config";
import { MikroORM } from "@mikro-orm/core";

module.exports = async () => {
  /**
   * This is to ensure that MikroORM cache is built before testing.
   */
  await MikroORM.init(config);
};
```
However, we get an error in running `await MikroORM.init()` in testing suites immediately after. This did not used to happen, and I am unsure if this is due to an update to Jest or to MikroORM. Removing the call to `MikroORM.init()` in the global setup resolves the issue.

**Stack trace**
```
Method Map.prototype.set called on incompatible receiver #<Map>
TypeError: Method Map.prototype.set called on incompatible receiver #<Map>
    at Map.set (<anonymous>)
    at C:\Users\{USER}\Project\node_modules\@mikro-orm\core\metadata\MetadataDiscovery.js:51:79
    at Array.forEach (<anonymous>)
    at MetadataDiscovery.processDiscoveredEntities (C:\Users\{USER}\Project\node_modules\@mikro-orm\core\metadata\MetadataDiscovery.js:51:44)
    at MetadataDiscovery.discover (C:\Users\{USER}\Project\node_modules\@mikro-orm\core\metadata\MetadataDiscovery.js:35:20)
    at MikroORM.discoverEntities (C:\Users\{USER}\Project\node_modules\@mikro-orm\core\MikroORM.js:105:25)
    at Function.init (C:\Users\{USER}\Project\node_modules\@mikro-orm\core\MikroORM.js:50:9)
    at Object.<anonymous> (C:\Users\{USER}\Project\src\database\foo\foo.service.spec.ts:7:11)
```

**Additional context**
[Here is a repo that can be used to duplicate the issue.
](https://github.com/kpervin/MikroORM-Request-Context-Issue)

**To Reproduce**
1. Run `src/database/foo/foo.service.spec.ts` in the provided repo

**Expected behavior**
Global setup should run fine, and create the `.ts` cache. Then, all tests should run.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.16.0 |
| typescript | 4.8.4 |
| mikro-orm | 5.5.3 |
| your-driver | MySQL |
| jest | 29.3.1 |
| ts-jest | 29.0.3 |
| @nestjs/testing | 9.2.0 |
| @mikro-orm/nest | 5.1.2 |

**Comment by B4nan** [comment]
Apparently, jest is doing some mocking of the `Map` primitive.

Pretty much the same issue in another library:

https://github.com/hapijs/joi/issues/2350

Will take a closer look later today, maybe there is something we can do, but this seems to be more of a jest issue.

**Reply by kpervin** [reply]
Yes, I saw that issue, but the fix is to explicitly call a mock function to remove automatic mocking, which here we don't want to mock. I recognize it's likely a Jest issue, but I thought it would be wise to post it here at the very least. I'll also open an issue on the Jest repo and link this issue.

**Comment by kpervin** [comment]
A potential workaround might be to add a `--tsNode` flag to the `generate:cache` command, as that's essentially all this call is being used for. It would be nice to have the CLI ability to generate cache not only for production use, but for TS use as well (such as for CI/CD). Is there any specific reason this was never added to begin with?

**Reply by B4nan** [reply]
The cache gets generated on first run in dev mode, I can't see the reason why you would want to warm that up (it sounds like it would rather slow it down, as you'd initialize the ORM twice). Warning up cache seems like a production only use case to me.

**Reply by kpervin** [reply]
It's due to that given the time it takes, it causes a timeout on tests that run for the default time of 5000ms. If the tests are run again directly after the first timeout, things run fine now that the cache has been created. Maybe I'm simply doing something weird in my testing setup? If so, let me know what I'm doing wrong, and I can help contribute to a `Testing` area of the documentation to help anyone else that may have run into this issue.

**Reply by kpervin** [reply]
It's also not an issue if you've run your application previously to generate the cache, or if you commit your cache after generating, but given the ever changing content of the cache in ongoing development, we were running into merge conflicts so we thought to omit it from our commit structure as it could be generated on a checkout hook or via the CLI at deployment.

**Reply by B4nan** [reply]
I see, I don't mind allowing this (the TS cache), PR welcome

**Reply by kpervin** [reply]
@B4nan [PR created](https://github.com/mikro-orm/mikro-orm/pull/3796)

**Comment by kpervin** [comment]
One thing I have not tried yet is [disabling `automock` in Jest](https://jestjs.io/docs/jest-object#jestdisableautomock). While it doesn't solve the need to cache `.ts` files before running tests, it may serve to remedy the issue of Jest attempting to mock the MikroORM module in this instance.

**Comment by darkbasic** [comment]
I'm weirdly getting this only when I run tests in vscode, but not when I run jest from the cmdline. Here is my config:

```
  "jest": {
    "globalSetup": "./tests/jestGlobalSetup.ts",
    "testEnvironment": "./tests/customEnvironment.ts",
    "testTimeout": 60000,
    "transform": {
      "^.+\\.(ts)$": [
        "ts-jest",
        {
          "tsconfig": "./tests/tsconfig.json"
        }
      ],
      "^.+\\.graphql$": "graphql-import-node/jest"
    }
  },
```

What could be the difference between the two?

**Comment by john-gom** [comment]
I dug into this a bit more and the problem seems to come when MikroORM clones the entity metadata after Jest (presumably) has mocked it. The clone method in @mikro-orm/core/utils/clone.js doesn't recognise that a Map is a Map and so clones it like a plain object.

I was able to workaround this by clearing the cached metadata in the global setup script so that it is re-created rather than using the cached (mocked) version. e.g.

```
import { MetadataStorage } from '@mikro-orm/core';

// Do whatever needs to be done with MikroORM
MetadataStorage.clear();
```

Obviously this doesn't help with the original problem of wanting to cache the metadata, but might help people who want to do other stuff in the global setup (in my case it was to start a testcontainer and run migrations).

**Reply by B4nan** [reply]
> I dug into this a bit more and the problem seems to come when MikroORM clones the entity metadata after Jest (presumably) has mocked it. The clone method in @mikro-orm/core/utils/clone.js doesn't recognise that a Map is a Map and so clones it like a plain object.

That's because it is no longer the same Map, jest is replacing the globals in the test context, see https://github.com/jestjs/jest/issues/2048 and https://github.com/jestjs/jest/issues/2549

So let's blame the actual villain 😃

**Reply by john-gom** [reply]
Sorry, no blame intended, just offering a workaround.

**Reply by B4nan** [reply]
Yeah, no worries. Btw I dont think you disable any caching with that call, its just the global decorator metadata storage, not the actual final metadata after discovery (those are cached).

**Reply by kallekulp** [reply]
So is this the current solution for that problem?

**Reply by drackp2m** [reply]
✅ I see it as a valid solution @kallekulp.

💚 Thanks for saving my life today @john-gom.

🚀 Your ORM is amazing @B4nan.
